kernel/src/time/wait.rs

```diff
@@ -5,48 +5,109 @@ use core::time::Duration;
 use ostd::sync::{WaitQueue, Waiter};
 
 use super::{clocks::JIFFIES_TIMER_MANAGER, timer::Timeout};
+use crate::prelude::*;
 
-/// A trait that provide the timeout related function for [`WaitQueue`]`.
+/// A trait that provide the timeout related function for [`Waiter`] and [`WaitQueue`]`.
 pub trait WaitTimeout {
     /// Wait until some condition returns `Some(_)`, or a given timeout is reached. If
-    /// the condition does not becomes `Some(_)` before the timeout is reached, the
-    /// function will return `None`.
-    fn wait_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Option<R>
+    /// the condition does not becomes `Some(_)` before the timeout is reached or `cancel_cond`
+    /// returns `Err`, this function will return corresponding `Err`.
+    fn wait_until_or_timeout<F, R, FCancel>(
+        &self,
+        cond: F,
+        cancel_cond: FCancel,
+        timeout: &Duration,
+    ) -> Result<R>
     where
```

Comment:
I hope `wait_until_or_timeout` should not accept `cancel_cond`. This can be done simply. You can rename current `wait_until_or_timeout` as `wait_until_or_timeout_cancelled` and add a new method named `wait_until_or_timeout` will call `wait_until_or_timeout_cancelled(....|| Ok(())...)`.

If user does not need to call `wait_until_or_timeout_cancelled` directly, it should be marked as `#[doc(hidden)]`